File: src/main/java/ca/warp7/pathplotter/state/PixelReference.java
package ca.warp7.pathplotter.state;

import edu.wpi.first.wpilibj.geometry.Translation2d;

public class PixelReference {

    private double pixelPerMetreWidth = 0;
    private double pixelPerMetreHeight = 0;
    private double originX = 0;
    private double originY = 0;

    public void set(
            FieldConfig config,
            double view_width,
            double view_height,
            double offsetX,
            double offsetY
    ) {
        var field_width_pixels = config.getRightPx() - config.getLeftPx();
        var field_height_pixels = config.getBottomPx() - config.getTopPx();
        var image_width = config.getImage().getWidth();
        var image_height = config.getImage().getHeight();

        var field_in_view_width = view_width / image_width * field_width_pixels;
        var field_in_view_height = view_height / image_height * field_height_pixels;

        pixelPerMetreWidth = field_in_view_width / config.getFieldLengthMetres();
        pixelPerMetreHeight = field_in_view_height / config.getFieldWidthMetres();

        var centerY = (config.getTopPx() + config.getBottomPx()) / 2.0;

        originX = offsetX + config.getLeftPx() / image_width * view_width;
        originY = offsetY + centerY / image_height * view_height;
    }

    public Translation2d scale(Translation2d point) {
        return new Translation2d(
                pixelPerMetreWidth * point.getX(),
                -pixelPerMetreHeight * point.getY()
        );
    }

    @SuppressWarnings("unused")
    public Translation2d inverseScale(Translation2d point) {
        return new Translation2d(
                -point.getX() / pixelPerMetreWidth,
                point.getY() / pixelPerMetreHeight
        );
    }

    public Translation2d transform(Translation2d point) {
        return new Translation2d(
                originX + pixelPerMetreWidth * point.getX(),
                originY - pixelPerMetreHeight * point.getY()
        );
    }

    public Translation2d inverseTransform(Translation2d point) {
        return new Translation2d(
                (point.getX() - originX) / pixelPerMetreWidth,
                -(point.getY() - originY) / pixelPerMetreHeight
        );
    }
}

File: src/main/kotlin/ca/warp7/planner2/state/PixelReference.kt
package ca.warp7.planner2.state

import edu.wpi.first.wpilibj.geometry.Rotation2d
import edu.wpi.first.wpilibj.geometry.Translation2d

/**
 * Defines functions to scale translations
 * to pixel sizes on the screen
 *
 * In JavaFX, x is to the right, y is down, origin at top-left
 *
 * In Path Planner coordinates, x is up, y is left,
 * origin is at the bottom-centre of the image
 *
 * The border is just extra. No calculations involved
 */
class PixelReference {

    private var pixelPerMetreWidth = 0.0
    private var pixelPerMetreHeight = 0.0

    private var originX = 0.0
    private var originY = 0.0

    fun set(
            width: Double,
            height: Double,
            offsetX: Double,
            offsetY: Double,
            fieldWidth: Double,
            fieldHeight: Double
    ) {
        pixelPerMetreWidth = width / fieldWidth
        pixelPerMetreHeight = height / fieldHeight

        originX = offsetX
        originY = offsetY + height / 2.0
    }

    fun scale(point: Translation2d) = Translation2d(
            pixelPerMetreWidth * point.x,
            -pixelPerMetreHeight * point.y
    )

    fun inverseScale(point: Translation2d) = Translation2d(
            -point.x / pixelPerMetreWidth,
            point.y / pixelPerMetreHeight
    )

    fun transform(point: Translation2d) = Translation2d(
            originX + pixelPerMetreWidth * point.x,
            originY - pixelPerMetreHeight * point.y
    )

    fun inverseTransform(point: Translation2d) = Translation2d(
            (point.x - originX) / pixelPerMetreWidth,
            -(point.y - originY) / pixelPerMetreHeight
    )
}